### PR TITLE
Fix negative indices in tracer

### PR DESCRIPTION
**Pull request by eellison**
Previously when tracing slicing & select negative indices would get normalized, fixing the index to the size of the traced tensor. This makes the behavior the same as script so aten::select with negative indices is emitted.

### CODE REVIEW

**Review by apaszke — APPROVED** [review]
Nice one!

File: test/test_jit.py [hunk 1]
```diff
@@ -4252,6 +4252,30 @@ def bar(x):
...
+        @torch.jit.trace(torch.zeros(10))
+        def foo_trace(x):
+            return x[-1:-3]
+
+        @torch.jit.script
+        def foo_script(x):
+            return x[-1:-3]
+
+        a = torch.arange(0, 8)
+        self.assertEqual(foo_script(a), foo_trace(a))
```
Comment (apaszke) [inline test/test_jit.py]: Actually the biggest problem with wrapping negative indices is that it specializes them to a particular size. Can you have a test which runs this function on two differently sized inputs and gets two differently sized outputs as a result?

**Review by apaszke — DISMISSED** [review]
Realized the tests are not strict enough

File: test/test_jit.py (hunk 1 above)
Comment (eellison) [reply test/test_jit.py]: The traced tensor and tensor that I test it on are a different size but I'll add another example.

**Review by zdevito — APPROVED** [review]
Seems good now.

File: torch/csrc/utils/python_compat.h [hunk 2]
```diff
@@ -2,6 +2,71 @@
...
 
+#if PY_VERSION_HEX < 0x03060100
+
+// PySlice_Unpack not introduced till python 3.6.1
+// included here for backwards compatibility
+// https://docs.python.org/3/c-api/slice.html#c.PySlice_Unpack
+// https://github.com/python/cpython/blob/master/Objects/sliceobject.c#L196
+
+//Replacing Py_BUILD_ASSERT
+#define BUILD_ASSERT(cond) \
```
Comment (apaszke) [inline torch/csrc/utils/python_compat.h]: It's a C++ codebase so you can just use `static_assert`.

**Review by apaszke — APPROVED** [review]
